### PR TITLE
clean up maintenance mode activation

### PR DESCRIPTION
**Pull request by george43g**
- clean up maintenance mode
- minor code cleanup/refactor

### CODE REVIEW

File: apps/backend-ops/src/users.ts [hunk 1]
```diff
@@ -192,20 +189,15 @@ export async function createUsers(): Promise<any> {
...
 
-  const users = await db.collection('users').get();
+  const usersBatch = getCollectionInBatches<User>(db.collection('users'), 'uid')
+  for await (const users of usersBatch) {
+    await runChunks(
+      users,
+      (user: User) => upsertWatermark(user, storageBucket, storage),
+      env?.['heavyChunkSize'] ?? 10
+    );
+  }
```
Comment (phpgeekfr) [inline apps/backend-ops/src/users.ts:200]: what is the purpose of this change?

File: libs/firebase-utils/src/lib/firebase-utils.ts [hunk 2]
```diff
@@ -53,14 +53,21 @@ export async function getDocAndPath(fullPath: string | undefined) {
...
+
+export async function runChunks<K, T extends (arg: K) => Promise<any>>(rows: K[], cb: T, rowsConcurrency?: number, verbose = true) {
   const chunks = chunk(rows, rowsConcurrency || env?.['chunkSize'] || 10);
+  const output: AsyncReturnType<T>[] = [];
   for (let i = 0; i < chunks.length; i++) {
     const c = chunks[i];
     if (verbose) { console.log(`Processing chunk ${i + 1}/${chunks.length}`); }
     const promises = c.map(cb);
-    await Promise.all(promises);
+    output.concat(await Promise.all(promises))
```
Comment (phpgeekfr) [inline libs/firebase-utils/src/lib/firebase-utils.ts:68]: missing `;` 
so runChunks is now typed ?

**Review by phpgeekfr — APPROVED** [review]

File: apps/backend-ops/src/users.ts (hunk 1 above)
Comment (phpgeekfr) [reply apps/backend-ops/src/users.ts:200]: With this, we have a new behavior in logs:
```
ssing chunk 84/93
Processing chunk 85/93
Processing chunk 86/93
Processing chunk 87/93
Processing chunk 88/93
Processing chunk 89/93
Processing chunk 90/93
Processing chunk 91/93
Processing chunk 92/93
Processing chunk 93/93
Processing chunk 1/93
Processing chunk 2/93
Processing chunk 3/93
Processing chunk 4/93
Processing chunk 5/93
Processing chunk 6/93
Processing chunk 7/93
Processing chunk 8/93
Processing chunk 9/93
Processing chunk 10/93
Processing chunk 11/93
```
because of batches,  count start from 1 again at each batch


(not an issue but just to let you know)

File: libs/firebase-utils/src/lib/firebase-utils.ts (hunk 2 above)
Comment (george43g) [reply libs/firebase-utils/src/lib/firebase-utils.ts:68]: yes!

Comment (phpgeekfr) [reply libs/firebase-utils/src/lib/firebase-utils.ts:68]: GJ

File: apps/backend-ops/src/users.ts (hunk 1 above)
Comment (george43g) [reply apps/backend-ops/src/users.ts:200]: I did this to future-proof it. Once it gets too big, we cant rely on a single collection().get() to get the entire collection, as previously we've seen that as the response gets really big, things go missing. This allows us to batch both the reads and the writes.

**Comment by phpgeekfr** [comment]
In fact, can you add setMaintenance in preparForTEsting also ? It will prevent ttriggers to be trigerred and will also prevent unwanted emails

**Comment by phpgeekfr** [comment]
> In fact, can you add setMaintenance in preparForTEsting also ? It will prevent ttriggers to be trigerred and will also prevent unwanted emails

and this will also be usefull when we run prepareForTesting on staging for example
